### ISSUE TITLE
Proposal: Create a universal encapsulator

### ISSUE THREAD

**Issue by aerbil313** [nushell/nushell#9864]
### Related problem

Writing our own versions of coreutils and such is not enough. If the idea of nushell is to go somewhere, structured-data tools need to be a complete ecosystem. Otherwise nushell will suffer on the matter of compatibility with string-data tools forever. Therefore I propose a universal encapsulator mechanism with a community-backed package repository.

## Description
```bash
$ encapsulate command args
```
1. Check the version of the command with a mechanism it supports (see below).
2. Download the latest version of the encapsulator package from the repository for the specific command version
3. Process structured data from stdin and args and feed it into the command according the the package
4. Take the command's output, convert it back to structured data according to the package and print it out.

To check program version:
* command -v or command --version
* $COMMAND_VERSION (if set)
* examine command's installation directory?

By default, nushell runs **all** commands with the encapsulator implicitly if they have an available encapsulator package in the repository. This is toggleable, e.g with an environment variable.

A shell builtin always runs commands directly without the encapsulator for temporary overriding. Example:
```bash 
$ cmd command # runs command directly
```

### Version pinning

Version pinning of encapsulator packages is allowed via env variables. When the user pins a package, latest version of the package which is backwards-compatible with the pinned version is downloaded.

Example pinning behavior:

1. The user sets $NUSHELL_PIN_NODE = 1.6 and types node in nushell.
2. Nushell will check the latest backwards-compatible package. Say it's version 1.92. But the package 1.92 is for node version 17, and the user has version 18.5 node command.
3. Nushell should error and tell the user that the pinned encapsulator package and the command are incompatible. It should also tell the user that he can either set $NUSHELL_NOPIN to true to ignore pinned packages or use the nopin builtin to ignore them for one command.

An environment variable can be set to use the exact pinned versions of all packages. This is useful for scripts.

Something like $NUSHELL_EXACT_PIN_NODE can be useful as well.

### Encapsulator packages

* Sem-versioned. Packages under version 1 are considered not production ready and nushell won't see them as available for encapsulation.
* An encapsulator package works with one specific version of a command.
* There can be multiple versions of an encapsulator package for one version of a command. All versions are stored in the repository.

## Problems

There are many programs which install commands with the same name. Packages in the repo can be given unique names, but when the user executes a command, it can be the output of many programs and nushell cannot know that. 

One possible solution is to to match known outputs of commands like their help texts (command --help) or examining their installation directories to determine which one is it, but this may prove cumbersome, especially for a community effort. 

### Nix integration

Integration with a universal package manager like Nix is a solution, but then the user will be able to use encapsulation only for packages he installed through Nix. Still I don't exactly see why that would be a problem, because Nix works everywhere and includes pretty much every program, and for those sliver of programs which are not on nixpkgs, the user can use them without encapsulation anyway. 

Nix integration can have those additional benefits:
1. Knowing the version of the command would be trivial, we can just check with nix.
2. We need not create a repository of our own. We can as well just use nixpkgs for encapsulation packages, if we are to depend on Nix for encapsulation.
3. There can be additional features like prompting the user to auto-install programs' versions which are compatible with the pinned encapsulation package versions. This can be very convenient for scripts, because I'd expect most serious scripts to pin their encapsulation packages.
4. Nushell scripts would be able to truly depend on just nushell, and will work everywhere Nushell exists. This is very powerful and convenient. If Nushell requires installing Nix as a dependency, scripts will be able to assume Nix is installed everywhere Nushell is, and thus be able to install their dependencies themselves safely. FYI, package installation with 'nix shell' command is temporary and doesn't alter or change the user's system in any way. Maybe Nushell can provide convenience functions or builtins for this, but this would probably be another proposal.

## Conclusion

With this proposal, Nushell will gain two new capabilites which no other shell I know of possesses:
1. True data-structure interoperability with external tools, even those not designed to use data structures
2. True portability including dependencies on external packages

I believe those together will ease one of the most important pain points of Nushell and skyrocket adoption.

**Comment by fdncred** [comment]
Sounds kind of similar to `jc` https://github.com/kellyjonbrazil/jc

**Comment by aerbil313** [comment]
Yeah, but this is for nu's data structures and has many other things as well, like a community-run parsers repo.

**Comment by sholderbach** [comment]
This would require a much larger actively maintained ecosystem, for now this seems outside of the immediate scope we want to achieve for Nushell's 1.0 release.

**Comment by aerbil313** [comment]
It seems reasonable. I think I should not close the issue, as you added post-1.0. Correct if I'm wrong.
